### ISSUE TITLE
[Data][Training] Optimize CPU bottlenecks in training data pipeline (iter_batches/iter_torch_batches)

### ISSUE THREAD

**Issue by OneSizeFitsQuorum** [ray-project/ray#63910]
# [Data][Training] Optimize CPU bottlenecks in training data pipeline (iter_batches/iter_torch_batches)

## Problem Statement

When using `iter_batches()` or `iter_torch_batches()` with `local_shuffle_buffer_size` for training workloads, there are several CPU-side bottlenecks that can limit throughput, especially in heterogeneous CPU/GPU setups where the GPU training loop is fast but data loading becomes the bottleneck.

While PR #62539 (incremental indexing optimization) significantly improved the `ShufflingBatcher` performance, there are still opportunities to optimize the training data pipeline and improve observability.

## Current Training Pipeline Architecture

The training data pipeline (`_internal/block_batching/iter_batches.py::_pipeline()`) consists of 6 stages:

1. **Prefetch blocks** - Fetch blocks from object store
2. **Resolve blocks** - `ray.get()` to materialize blocks
3. **blocks_to_batches** - Batch construction + local shuffle (via `ShufflingBatcher`)
4. **Format + collate** - Convert to target format (numpy/torch) in threadpool
5. **Finalize** - Move to GPU (if applicable)
6. **Restore order** - Reorder batches after threadpool operations

### 1. **ShufflingBatcher Observability Gap**

**Problem**: Even with the incremental index optimization (#62539), there are no Prometheus metrics to monitor:
- Number of compactions triggered
- Compaction latency (per compaction event)
- Buffer size vs threshold
- Whether shuffle buffer is the actual bottleneck

**Impact**: Users cannot determine if local shuffle is their bottleneck without manual profiling. This was noted in #42317 but remains unaddressed.

**Proposed Metrics**:
- `ray_data_shuffle_buffer_compaction_count` (counter)
- `ray_data_shuffle_buffer_compaction_latency_ms` (histogram)
- `ray_data_shuffle_buffer_size_rows` (gauge)
- `ray_data_shuffle_buffer_size_bytes` (gauge)

### 2. **Prefetch Strategy Limitations**

**Problem**: Current prefetch implementation has several limitations:
- Synchronous prefetch can block the training loop
- No distinction between local vs remote blocks (network latency ignored)
- Static prefetch window doesn't adapt to GPU utilization
- No metrics to observe prefetch effectiveness

**Code Reference**: `_internal/block_batching/iter_batches.py:410-420`

```python
while current_window_size < num_rows_to_prefetch:
    next_ref_bundle = get_next_ref_bundle()
    sliding_window.extend(next_ref_bundle.blocks)
```

**Potential Optimizations**:
- Asynchronous prefetch with background thread
- Adaptive prefetch size based on GPU utilization
- Separate queues for local vs remote blocks
- Prefetch metrics: `ray_data_prefetch_queue_depth`, `ray_data_prefetch_latency_ms`

### 3. **ThreadPool Saturation**

**Problem**: The format/collate threadpool uses a static size:
```python
num_threadpool_workers = min(4, prefetch_batches)
```

This doesn't account for:
- Available CPU cores
- Complexity of `collate_fn`
- Actual threadpool utilization

**Impact**: Under-utilization (wasted CPU) or over-utilization (contention) depending on workload.

**Potential Optimizations**:
- Dynamic threadpool sizing based on CPU cores
- ThreadPool utilization metrics
- Separate threadpool for format vs collate (if collate is expensive)

### 4. **Block Resolution Overhead**

**Problem**: `ray.get()` is called synchronously for each block:
```python
def _resolve_block_refs(...):
    return resolve_block_refs(block_ref_iter, stats)
```

For workloads with many small blocks, the overhead of individual `ray.get()` calls becomes significant.

**Potential Optimizations**:
- Batch `ray.get()` calls (get multiple blocks at once)
- Overlap resolution with prefetching
- Async resolution with `ray.wait()`

## Reproduction Scenario

Typical training workload that hits these bottlenecks:
```python
import ray
from ray.train.torch import TorchTrainer

def train_fn():
    train_ds = ray.data.read_parquet("s3://large-dataset/")
    train_loader = train_ds.iter_torch_batches(
        batch_size=256,
        local_shuffle_buffer_size=10000,
        prefetch_batches=2
    )
    
    for batch in train_loader:
        # Fast GPU training (e.g., 5ms per batch)
        # But data loading takes 15ms per batch
        model.train_step(batch)
```

**Observed behavior** (example):
- GPU utilization: 30% (starved for data)
- CPU utilization: 95% (bottlenecked on data loading)
- No metrics to identify which stage is the bottleneck

## Proposed Solution

I propose a phased approach:

### Phase 1: Observability (Small PR, ~150 lines)
Add Prometheus metrics for:
- `ShufflingBatcher` compaction events
- Prefetch queue depth and latency
- ThreadPool utilization

This will help users identify their specific bottleneck and guide future optimizations.

### Phase 2: Prefetch Optimization (Medium PR, ~300 lines)
- Implement adaptive prefetch based on GPU utilization
- Add async prefetch mode
- Separate local vs remote block handling

### Phase 3: ThreadPool Optimization (Medium PR, ~200 lines)
- Dynamic threadpool sizing
- Add utilization metrics
- Profile-guided tuning recommendations

## Questions for Maintainers

1. **Priority**: Is this the right direction for training data pipeline optimization? Or are there other bottlenecks I should focus on first?

2. **Metrics Design**: Are the proposed metric names aligned with Ray's Prometheus conventions? Should they be under `ray_data_*` or a more specific namespace like `ray_data_training_*`?

3. **Phase 1 Scope**: Would the maintainers accept a focused PR on observability first, or do you prefer a more comprehensive solution that includes optimizations?

4. **Related Work**: Are there any ongoing efforts in this area that I should coordinate with?

## Additional Context

- Related Issue: #42317 (Local shuffle buffer throughput regression - now closed)
- Related PR: #62539 (Incremental indexing optimization for ShufflingBatcher - merged)
- Related PR: #63744 (OTel Prometheus label normalization - merged)

## Environment

- Ray version: 2.55.1 (will be in 2.56.0+)
- Python version: 3.11+
- OS: Linux/macOS

---

**Next Steps**: Based on community feedback, I'll start with Phase 1 (observability) and iterate from there.
